docs/Fleet management Design.md

```diff
@@ -101,52 +125,122 @@ type ResourceList map[ResourceName]resource.Quantity
+    Message:               Managed cluster is available
+    Reason:                ManagedClusterAvailable
+    Status:                True
+    Type:                  ManagedClusterConditionAvailable
+  Version:
+    Kubernetes:  v1.20.9
+```
+
 The last one has to be provided by the member cluster. There is already a customer object named `ClusterClaim` that contains
-the required information.  Therefore, we propose that the `managedCluster` controller to just list the `ClusterClaim` objects in the 
-member cluster to fill this part.
+the [required information] (https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/4-cluster-claims). 
+Each claim contains just one of the required property. Therefore, the hub controller can just list the `ClusterClaim` objects in the
+member cluster and fill the `ClusterClaim` fields in the status.
 
-#### Work with `work` API
+### Work with `work` API
 The `work` API is the way how OCM manage "applications". Its controller is running in the member cluster in the current pull
 model. To make it work with push model, we need to implement it with the `managedCluster` controller. Here is how it works
 
 * The `managedCluster` controller reconciler keeps an in-memory map of member cluster -> manifestwork controller
-* for each new member cluster, it creates a new manifestwork controller that watches all the `ManifestWork` objects in the 
-corresponding namespace and run it in separate go routine.
+* for each new member cluster, it creates a new manifestwork sub-controller that watches all the `ManifestWork` objects in the 
+corresponding namespace and run it in a separate go routine.
+* The manifestwork sub-controller functions pretty much like the existing worker controller groups in the member cluster which
+  * for each `ManifestWork` object in the namespace, it extracts the content of the objects and apply them to the corresponding member cluster
+  * it merges the status of the applied object in the member cluster back to the `ManifestWork` object in the hub cluster.
+  * It uses a finalizer to make sure that when a `ManifestWork` object is deleted from the hub cluster, all the resources used in the member cluster is also removed. 
 
 
-#### Remove a member cluster from the fleet
+### Remove a member cluster from the fleet
 We will add a finalizer in the `managedCluster` object to make sure that we will clean up the resources in the corresponding
 member cluster when one removes the object itself. Potential things to clean up
 
-1. Any in memory object associated with the member cluster such as the worker reconcile loop
-2. Fleet related Claims/Secrete/configMap in the member cluster
-3. Any metadata recording on the side (i.e a database)
+1. The lease configMap in the member cluster
+2. Any in memory object associated with the member cluster such as the worker reconcile loop
+3. Fleet related Claims/Secrete/configMap in the member cluster
 
```

Comment:
Yes, we should discuss that. I haven't looked into the addon mechanism too much.  IIRC, the registration agent dumps a cert in a predefined secret for the addons to consume. It seem then we can run those in the hub too as long as they have access to both clusters

Reply:
some addon cannot, for instance, you may want to run a networking addon that may need to set some cni in the spoke

Reply:
I wonder if there is any specific reason that cni can't be setup using a controller that has admin access to its api-server?

Reply:
not that we envision we will need to set up CNI on the nodes

Reply:
what most of the addon did today is they define (on the hub) what they want to deploy on the managed cluster, and how they want to connect to the hub or other central endpoint. For registration part, It make sense that cluster registration is not needed, however addon registration is need still since otherwise addon cannot know how to connect to the hub from the managedcluster.

A simple example in submariner is: submariner has a gateway pod to setup VPN tunnel among managed clusters. This gateway pod needs to get info from hub cluster, and it cannot run on hub cluster. The addon controller of submariner will deploy the pod and also tell registration to setup the credential for this pod to talk to hub.

Reply:
I see. Yeah, looks like the addon model you described needs to run on the member cluster. The `push` model definitely won't solve that.  Just to be clear, we don't intend to use `push` model as a replacement of the current OCM pull model at all. We view it as a lightweight version to get things started fast in the open source community. We envision all the enterprise users will use pull model in production.

Reply:
yeah, I think push model is valuable even in some production environment.